graph/path/shortest.go

```diff
@@ -420,38 +420,67 @@ func (p ShortestAlts) AllTo(vid int64) (paths [][]graph.Node, weight float64) {
 	}
 
 	seen := make([]bool, len(p.nodes))
-	paths = p.allTo(from, to, seen, []graph.Node{p.nodes[to]}, nil)
+	p.allTo(from, to, seen, []graph.Node{p.nodes[to]}, func(path []graph.Node) {
+		paths = append(paths, append([]graph.Node(nil), path...))
+	})
 	weight = p.dist[to]
 
 	return paths, weight
 }
 
+// AllToFunc calls fn all shortest paths to v. Paths containing zero-weight
+// cycles are not considered. If a negative cycle exists between u and v, no
+// path is considered. The fn closure must not retain the path parameter.
+func (p ShortestAlts) AllToFunc(vid int64, fn func(path []graph.Node)) {
+	from := p.indexOf[p.from.ID()]
+	to, toOK := p.indexOf[vid]
+	if !toOK || len(p.next[to]) == 0 {
+		if p.from.ID() == vid {
+			fn([]graph.Node{p.nodes[from]})
+		}
+		return
+	}
+
+	_, weight, unique := p.To(vid)
+	if math.IsInf(weight, -1) && !unique {
+		return
+	}
+
+	seen := make([]bool, len(p.nodes))
+	p.allTo(from, to, seen, []graph.Node{p.nodes[to]}, fn)
+}
+
 // allTo recursively constructs a slice of paths extending from the node
 // indexed into p.nodes by from to the node indexed by to. len(seen) must match
 // the number of nodes held by the receiver. The path parameter is the current
-// working path and the results are written into paths.
-func (p ShortestAlts) allTo(from, to int, seen []bool, path []graph.Node, paths [][]graph.Node) [][]graph.Node {
+// working path and the results passed to fn.
+func (p ShortestAlts) allTo(from, to int, seen []bool, path []graph.Node, fn func(path []graph.Node)) {
 	seen[to] = true
 	if from == to {
 		if path == nil {
-			return paths
+			return
 		}
 		ordered.Reverse(path)
-		return append(paths, path)
+		fn(path)
+		ordered.Reverse(path)
+		return
```

Comment:
isn't there a possible change of semantics here ?
before the PR, we (seemed to) reverse the order of nodes in place and then copy inside of `paths`.
but now, we (probably) copy into a user-owned `paths` the original order of nodes, and then reverse the order of nodes of `path`, but AFAICT, the order of nodes in the user-owned `paths` isn't altered. is it ?

do we test for this `from==to` case in the set of tests ?

Reply:
We have a partial path here, so we show it to the caller in the order they are expecting it, allow them to look at it, and then put it back to how we need it to continue working with it. This is why they may not hold it after they return from their closure.